### ISSUE TITLE
ComplexWarning in VJP when using complex matrix multiplication

### ISSUE THREAD

**Issue by mathisgerdes** [google/jax#21188]
### Description

Consider a function $f(a) = a b$ with $b$ some complex parameter and $a$ real. I want to compute the vector-Jacobian product $t^* \partial f(a)$. When $a b$ represents matrix multiplication, a warning is raised about an imaginary part being discarded.

For $a$, $b$ scalar it works as expected and without warning (imported `jax`, `jax.numpy as jnp`, and `warnings`):
```python
a = jnp.array(1.0)
b = jnp.array(1.0j)
t = jnp.array(1.0j)

with warnings.catch_warnings():
  warnings.simplefilter("error", jnp.ComplexWarning)
  
  _, vjp_fn = jax.vjp(lambda x: x * b, a)
  vjp_fn(t)
```

However, if we promote $a$ and $b$ to matrices, this raises a ComplexWarning:
```python
a = jnp.array([[1.0]])
b = jnp.array([[1.0j]])
t = jnp.array([[1.0j]])

with warnings.catch_warnings():
  warnings.simplefilter("error", jnp.ComplexWarning)
  
  _, vjp_fn = jax.vjp(lambda x: x @ b, a)
  vjp_fn(t)
```

```
ComplexWarning: Casting complex values to real discards the imaginary part
```

I would not expect a complex warning in the latter case. Although the output is the same in both cases (up to shapes), the warning suggests something could go wrong. Unless I'm missing something, the two behaviors are at least inconsistent. 

The warning is not raised when $a$ is a matrix but $b$ is not (matrix-scalar multiplication), but it also occurs when using `einsum`.

### System info (python version, jaxlib version, accelerator, etc.)

Tested on CPU:

```
jax:    0.4.23
jaxlib: 0.4.23
numpy:  1.26.3
python: 3.9.9 (v3.9.9:ccb0e6a345, Nov 15 2021, 13:06:05)  [Clang 13.0.0 (clang-1300.0.29.3)]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
```

as well as on Colab (jax `0.4.26` both on CPU and GPU).

**Comment by mattjj** [comment]
Thanks for the clear report! I'll take a look, though my initial guess is that this is a spurious warning we should suppress.

**Comment by mathisgerdes** [comment]
Yes, that would be a good resolution. I'm also unaware of anything going wrong. But the warning made me waste some time searching for what I might be doing wrong, so I thought fixing this might be useful.

**Comment by ScottRLawrence** [comment]
I believe that the following is the same issue, although it's difficult to be certain. Let me know if you'd like a separate report.

The following code gives a `ComplexWarning` only after printing `TWO`, even though `f1` and `f2` are functionally the same. The `ComplexWarning` appears to have its origins in the interaction of `vjp` with `dot_general`, and if the functionality of `dot_general` is implemented by hand, the warning doesn't appear.

```python
import jax
import jax.numpy as jnp

def f1(x):
    return jnp.sum((x+0j)*x)

def f2(x):
    return jnp.dot(x+0j, x)

x = jnp.zeros(3, dtype=jnp.float32)
v = jnp.ones((), dtype=jnp.complex64)

print('ONE')
jax.vjp(f1, x)[1](v)
print('TWO')
jax.vjp(f2, x)[1](v)
```

(I initially encountered this by using `einsum`, and only after printing out a jaxpr or two did I get the idea to try with `dot`.)
